src/transformers/utils/import_utils.py

```diff
@@ -1581,3 +1655,173 @@ def direct_transformers_import(path: str, file="__init__.py") -> ModuleType:
+      to execute correctly without instantiating it
+    - The '@register' string is used to dynamically import objects
+    """
+
+    if not isinstance(backends, tuple):
+        raise ValueError("Backends should be a tuple.")
+
+    def inner_fn(fun):
+        fun.__backends = backends
+        return fun
+
+    return inner_fn
+
+
+@lru_cache()
+def define_import_structure(module_path):
+    import_structure = {}
+    if os.path.isdir(module_path):
+        for f in os.listdir(module_path):
+            if f != "__pycache__" and os.path.isdir(os.path.join(module_path, f)):
+                import_structure[f] = define_import_structure(os.path.join(module_path, f))
+        directory = module_path
+    else:
+        directory = os.path.dirname(module_path)
+
+    adjacent_modules = [f for f in os.listdir(directory) if not os.path.isdir(os.path.join(directory, f))]
+
+    if "__init__.py" in adjacent_modules:
+        adjacent_modules.remove("__init__.py")
+
+    module_requirements = {}
+    for module_name in adjacent_modules:
+        with open(os.path.join(directory, module_name)) as f:
+            file_content = f.read()
+
+        # Remove the .py suffix
+        if module_name.endswith(".py"):
+            module_name = module_name[:-3]
+
+        previous_line = ""
+        previous_index = 0
+
+        lines = file_content.split("\n")
+        for index, line in enumerate(lines):
+            # This allows registering items with other decorators. We'll take a look
+            # at the line that follows at the same indentation level.
+            if line.startswith((" ", "\t", "@", ")")) and not line.startswith("@register"):
+                continue
+
+            # Skipping line enables putting whatever we want between the
+            # register() call and the actuall class/method definition.
+            # This is what enables having # Copied from statements, docs, etc.
+            skip_line = False
+
+            if "@register" in previous_line:
+                skip_line = False
+
+                # Backends are defined on the same line as register
+                if "backends" in previous_line:
+                    backends_string = previous_line.split("backends=")[1].split("(")[1].split(")")[0]
+                    backends = tuple(sorted([b.strip("'\",") for b in backends_string.split(", ")]))
+
+                # Backends are defined in the lines following register, for example such as:
+                # @register(
+                #     backends=(
+                #             "sentencepiece",
+                #             "torch",
+                #             "tf",
+                #     )
+                # )
+                #
+                # or
+                #
+                # @register(
+                #     backends=(
+                #             "sentencepiece", "tf"
+                #     )
+                # )
+                elif "backends" in lines[previous_index + 1]:
```

Comment:
I think in this specifc case, parsing the super small code would be simpler with AST, WDYT? 
not parsing the entire module, but just a few lines might be fast enough!

Reply:
or regex might be a bit simpler ?

Reply:
I haven't done this yet. If we see this doesn't affect performance I'm happy to go with that in a future PR.